src/hotspot/cpu/ppc/templateTable_ppc_64.cpp

```diff
@@ -3509,36 +3509,24 @@ void TemplateTable::fast_invokevfinal(int byte_no) {
   assert(byte_no == f2_byte, "use this argument");
   Register Rcache  = R31;
   __ load_method_entry(Rcache, R11_scratch1);
-  invokevfinal_helper(Rcache, R11_scratch1, R12_scratch2, R22_tmp2);
+  invokevfinal_helper(Rcache, R11_scratch1, R12_scratch2, R22_tmp2, R23_tmp3);
 }
 
-void TemplateTable::invokevfinal_helper(Register Rcache, Register Rscratch1, Register Rscratch2, Register Rscratch3) {
+void TemplateTable::invokevfinal_helper(Register Rcache,
+                                        Register Rscratch1, Register Rscratch2, Register Rscratch3, Register Rscratch4) {
 
-  assert_different_registers(Rcache, Rscratch1, Rscratch2, Rscratch3);
+  assert_different_registers(Rcache, Rscratch1, Rscratch2, Rscratch3, Rscratch4);
 
-  // Load receiver from stack slot.
-  Register Rmethod = Rscratch3;
-  __ ld(Rmethod, in_bytes(ResolvedMethodEntry::method_offset()), Rcache);
+  Register Rrecv     = Rscratch2,
+           Rmethod   = Rscratch3,
+           Rret_addr = Rscratch4;
+  prepare_invoke(Rcache, Rret_addr, Rrecv, Rscratch1);
 
-  // Get return address.
-  Register Rtable_addr = Rscratch2,
-           Rret_addr   = Rcache,
-           Rret_type   = Rscratch1;
-  // Get return type. It's coded into the upper 4 bits of the lower half of the 64 bit value.
-  __ lbz(Rret_type, in_bytes(ResolvedMethodEntry::type_offset()), Rcache);
-  __ load_dispatch_table(Rtable_addr, Interpreter::invoke_return_entry_table());
-  __ sldi(Rret_type, Rret_type, LogBytesPerWord);
-  __ ldx(Rret_addr, Rret_type, Rtable_addr); // kills Rcache
-
-  Register Rnum_params = Rscratch2,
-           Rrecv       = Rscratch2;
-  __ ld(Rnum_params, in_bytes(Method::const_offset()), Rmethod);
-  __ lhz(Rnum_params /* number of params */, in_bytes(ConstMethod::size_of_parameters_offset()), Rnum_params);
```

Comment:
[`ConstMethod::_size_of_parameters`](https://github.com/openjdk/jdk/blob/0712b22a3ae7075304e5925365429e1d85bd173c/src/hotspot/share/oops/constMethod.hpp#L208) is the size of the parameter block in words. `prepare_invoke` uses `ResolvedMethodEntry::_number_of_parameters` which is `Number of arguments for method`. I'd expect the location of the receiver to depend on the size of the parameters and not their number. How does this work?

Reply:
One is a copy of the other. See usages of `method_entry->fill_in` in src/hotspot/share/oops/cpCache.cpp. E.g.
`method_entry->fill_in((u1)as_TosState(method->result_type()), (u2)method()->size_of_parameters());`
Scaling happens in `load_receiver`.

Reply:
Thanks for looking it up for me :)
So `ResolvedMethodEntry::_number_of_parameters` is a misnomer or am I missing something?

Reply:
So, yes, the name doesn't tell what it actually is. It's the number of slots and not the number of arguments. But that's out of scope for this PR.

Reply:
It is out of this scope. And it is a misnomer that will confuse people.

Reply:
Agreed. Thanks for the review!

Reply:
@matias9927: `ResolvedMethodEntry::_number_of_parameters` is a misleading name. It's not the number of parameters, it's the number of parameter stack slots. Not every parameter uses only one slot. You may want to improve this in the future.